server/src/main/java/org/elasticsearch/persistent/PersistentTasksClusterService.java

```diff
@@ -243,19 +268,42 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
     public void clusterChanged(ClusterChangedEvent event) {
         if (event.localNodeMaster()) {
             if (shouldReassignPersistentTasks(event)) {
-                logger.trace("checking task reassignment for cluster state {}", event.state().getVersion());
-                clusterService.submitStateUpdateTask("reassign persistent tasks", new ClusterStateUpdateTask() {
-                    @Override
-                    public ClusterState execute(ClusterState currentState) {
-                        return reassignTasks(currentState);
-                    }
+                // We want to avoid a periodic check duplicating this work
+                periodicRechecker.cancel();
+                reassignPersistentTasks(event.state().getVersion());
+            } else {
+                scheduleRecheckIfUnassignedTasks(event.state());
+            }
+        }
+    }
 
-                    @Override
-                    public void onFailure(String source, Exception e) {
-                        logger.warn("failed to reassign persistent tasks", e);
-                    }
-                });
+    /**
+     * Submit a cluster state update to reassign any persistent tasks that need reassigning
+     */
+    private void reassignPersistentTasks(long currentStateVersion) {
+        logger.trace("checking task reassignment for cluster state {}", currentStateVersion);
```

Comment:
it's weird to have a parameter passed in for just a version log under trace.  Can you remove it and leave the logging in the event method?